CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt

```diff
@@ -6,39 +6,140 @@ import java.io.File
+        // (demo or staging) has higher priority.
+        //
+        // According to:
+        //
+        // https://developer.android.com/studio/build/build-variants#flavor-dimensions
+        //
+        // "... product flavors belonging to higher-priority flavor dimension appear first,
+        // followed by those from lower-priority dimensions, followed by the build type."
+        //
+        // It means that `staging` flavor has higher priority than `demo` flavor, because it appears first
+        // in variant's name: `stagingDemoDebug`. Moreover, more complex flavors has higher priority than less complex,
+        // so `stagingDemo` flavor has higher priority than either `staging` or `demo` flavors.
+        //
+        // Someone could ask a question: "Why `staging` flavor is first in variant's name `stagingDemoDebug`?
+        // Why the variant name is `stagingDemoDebug` not `demoStagingDebug`?" The name (and flavors' priority) is determined by
+        // order of flavorDimensions in build.gradle file. In our case this is:
+        //
+        //        flavorDimensions "api", "mode"
+        //
+        // `staging` belongs to `api` dimension, so it's the first
+        // `demo` belongs to `mode` dimesion, so it's the second in `stagingDemoDebug` variant's name
+        //
+        // Finally, resource merging priority is following:
+        // main < demo < staging < demoStaging < debug < stagingDemoDebug
+        //
+        // It corresponds to Cockpit files priority:
+        //
+        // cockpit.yml < cockpitDemo.yml < cockpitStaging.yml < cockpitDemoStaging.yml < cockpitDebug.yml < cockpitStagingDemoDebug.yml
+        //
+        // Let's find out how to create that list of files:
+        // [cockpit.yml, cockpitDemo.yml, cockpitStaging.yml, cockpitDemoStaging.yml, cockpitDebug.yml, cockpitStagingDemoDebug.yml]
+        //
 
-        val ordered = ArrayList<String>()
-        ordered.add("") // main
-        val singleFlavors = dimensions.reversed().map { dimension ->
+        val variantsOrderedFromLowestPriority = ArrayList<String>()
+
+        // `main` has the lowest priority, so we add it as a first
+        // Empty string corresponds with `main` variant
+        // because cockpit.yml doesn't contains "main" string in its name
+        variantsOrderedFromLowestPriority.add("")
+
+        // Next, the lowest priority has `demo` and `staging` flavors.
+        //
+        // Declared dimension's order is following:
+        //
+        //        flavorDimensions "api", "mode"
+        //
+        // so we have to reverse dimensions list, to add `demo` (from `mode` dimension) first
+        // and `staging` (from `api` dimension) after `demo`
+        //
+        // Only `staging` and `demo` flavors should be considered, because only those two flavors appear in
+        // variant `stagingDemoDebug`. So we find only those variants which appears in variant's name.
```

Comment:
"So we find only those variants which appears in variant's name." – I guess you meant "So we find only those flavors which appear in variant's name". :)